pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```diff
@@ -86,11 +92,161 @@ type Interface interface {
+// will be initialized with DriverName, ProxyInTreeVolumePlugin, and Available.
+func (nim *nodeInfoManager) InitializeMigratedDrivers() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) {
+		return fmt.Errorf("initializeMigratedDrivers called but CSINodeInfo or CSIMigration is disabled")
+	}
+
+	var updateErrs []error
+
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	// TODO(dyzz): This should probably retry forever!!
+	// TODO(TODO): Change this into reconcile pattern
+	// Loop to initialize Specs
+	err := wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
+		nodeInfo, err := csiKubeClient.CsiV1alpha1().CSINodeInfos().Get(string(nim.nodeName), metav1.GetOptions{})
+		if nodeInfo == nil || errors.IsNotFound(err) {
+			nodeInfo, err = nim.CreateCSINodeInfo()
+		}
+		if err != nil {
+			updateErrs = append(updateErrs, err)
+			return false, nil
+		}
+		err = nim.tryInitializeMigratedDriverSpecs(nodeInfo)
+		if err != nil {
+			updateErrs = append(updateErrs, err)
+			return false, nil
+		}
+		return true, nil
+
+	})
+
+	// Loop to initialize Status
+	// TODO(TODO) We can remove this once nodeinfomanager is refactored into a reconcile pattern because
+	// then we can just add the specs in the above loop and the reconciler will fill in the statuses based
+	// on that.
+	err = wait.ExponentialBackoff(updateBackoff, func() (bool, error) {
+		nodeInfo, err := csiKubeClient.CsiV1alpha1().CSINodeInfos().Get(string(nim.nodeName), metav1.GetOptions{})
+		if err != nil {
+			updateErrs = append(updateErrs, err)
+			return false, nil
+		}
+		err = nim.tryInitializeMigratedDriverStatuses(nodeInfo)
+		if err != nil {
+			updateErrs = append(updateErrs, err)
+			return false, nil
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("error updating node: %v; caused by: %v", err, utilerrors.NewAggregate(updateErrs))
+	}
+	return nil
+}
+
+func (nim *nodeInfoManager) tryInitializeMigratedDriverSpecs(nodeInfo *csiv1alpha1.CSINodeInfo) error {
+	csiKubeClient := nim.volumeHost.GetCSIClient()
+	if csiKubeClient == nil {
+		return fmt.Errorf("error getting CSI client")
+	}
+
+	existingDrivers := sets.String{}
+
+	// Do existing updates
+	// Clone driver list, omitting the driver that matches the given driverName
+	newDriverSpecs := []csiv1alpha1.CSIDriverInfoSpec{}
+	for _, driverInfoSpec := range nodeInfo.Spec.Drivers {
+		// No need to change any specs
+		newDriverSpecs = append(newDriverSpecs, driverInfoSpec)
+		existingDrivers.Insert(driverInfoSpec.Name)
+	}
+
+	for driver, _ := range nim.migratedDrivers {
+		if existingDrivers.Has(driver) {
+			// Skip adding new driver because already updated previously
+			continue
+		}
+
+		driverSpec := csiv1alpha1.CSIDriverInfoSpec{
+			Name:         driver,
+			NodeID:       "",
```

Comment:
what are observers told to do with an empty value? does that mean "this is not yet populated" or "use the node name as the node ID"?

Reply:
It means "this is not yet populated". At the time of initialization if the driver does not exist we have no way of getting that information.

Reply:
that's fine, as long as everything making use of this field has the same understanding, and waits for the field to be populated rather than assuming an empty field means to use some other value as a fallback